src/main/java/io/cryostat/rules/MatchExpressionManager.java

```diff
@@ -0,0 +1,190 @@
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.rules;
+
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.script.ScriptException;
+
+import io.cryostat.core.log.Logger;
+import io.cryostat.platform.PlatformClient;
+import io.cryostat.platform.ServiceRef;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+import dagger.Lazy;
+
+public class MatchExpressionManager {
+    private final MatchExpressionValidator matchExpressionValidator;
+    private final Lazy<MatchExpressionEvaluator> matchExpressionEvaluator;
+    private final PlatformClient platformClient;
+    private final MatchExpressionDao dao;
+    private final Gson gson;
+    private final Logger logger;
+
+    MatchExpressionManager(
+            MatchExpressionValidator matchExpressionValidator,
+            Lazy<MatchExpressionEvaluator> matchExpressionEvaluator,
+            PlatformClient platformClient,
+            MatchExpressionDao dao,
+            Gson gson,
+            Logger logger) {
+        this.matchExpressionValidator = matchExpressionValidator;
+        this.matchExpressionEvaluator = matchExpressionEvaluator;
+        this.platformClient = platformClient;
+        this.dao = dao;
+        this.gson = gson;
+        this.logger = logger;
+    }
+
+    public int addMatchExpression(String matchExpression)
+            throws MatchExpressionValidationException {
+        matchExpressionValidator.validate(matchExpression);
+        MatchExpression expression = dao.save(new MatchExpression(matchExpression));
+        return expression.getId();
+    }
+
+    public Optional<MatchExpression> get(int id) {
+        return dao.get(id);
+    }
+
+    public List<MatchExpression> getAll() {
+        return dao.getAll();
+    }
+
+    public boolean delete(int id) {
+        return dao.delete(id);
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(int id) {
+        Optional<MatchExpression> matchExpression = dao.get(id);
+        if (matchExpression.isEmpty()) {
+            return Set.of();
+        }
+        return resolveMatchingTargets(matchExpression.get().getMatchExpression());
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(String matchExpression) {
```

Comment:
I think both this method and the one above could be replaced by a `resolveMatchingTargets(MatchExpression expr)`. This puts the onus on the caller to first call `get(id)` first or do some other query, but this way the manager isn't accepting/evaluating arbitrary expressions (that's what the MatchExpressionEvaluator is for) and the query potentially coming up with no results only needs to be handled in one place.

Reply:
I may still need to have a method that accepts a raw string though because of the dry run POST handler without adding the match expression to the backend. The raw string can still have be evaluated by the `MathExpressionEvaluator` , but it bypasses the need to add it to the backend.